File: frontend/did.js
"use strict";
/*
 * did:key in the browser — the exact mirror of crypto/did.py.
 *
 * The browser generates and holds Ed25519 keys already (see app.js); this file
 * only *renders* one of those public keys as a W3C did:key identifier, so the
 * UI can show a portable, self-certifying name for the identity instead of a
 * bare 64-character hex blob. No new key material, no signing, nothing sent to
 * the node: the hex public key remains what gets put in a transaction's
 * `sender`.
 *
 * Encoding (must stay byte-for-byte identical to the Python side, which is
 * what tests/test_did_js.py asserts):
 *
 *   did:key:z<base58btc( 0xed 0x01 || 32 raw public key bytes )>
 *
 * where 0xed 0x01 is the ed25519-pub multicodec varint and "z" is multibase
 * base58btc.
 */

(function (global) {
  const ED25519_MULTICODEC_PREFIX = [0xed, 0x01];
  const MULTIBASE_BASE58BTC = "z";
  const DID_KEY_PREFIX = "did:key:";
  const PUBLIC_KEY_LEN = 32;

  // Bitcoin's base58 alphabet — no 0, O, I or l, the characters readers confuse.
  const B58_ALPHABET =
    "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";

  /*
   * base58btc encode. JS has no big integers in Uint8Array form, and the keys
   * here are far past Number's exact range, so this does long division on the
   * byte array itself (repeatedly divide the whole buffer by 58, collecting
   * remainders) — the standard byte-wise algorithm, exact for any length.
   * Leading zero bytes carry no weight in that division and are re-emitted as
   * "1", which is what makes the encoding reversible.
   */
  function b58encode(bytes) {
    let zeros = 0;
    while (zeros < bytes.length && bytes[zeros] === 0) zeros++;

    const digits = [];
    const buf = Array.from(bytes.slice(zeros));
    let start = 0;
    while (start < buf.length) {
      let remainder = 0;
      for (let i = start; i < buf.length; i++) {
        const acc = remainder * 256 + buf[i];
        buf[i] = Math.floor(acc / 58);
        remainder = acc % 58;
      }
      digits.push(remainder);
      while (start < buf.length && buf[start] === 0) start++;
    }

    return (
      "1".repeat(zeros) +
      digits
        .reverse()
        .map((d) => B58_ALPHABET[d])
        .join("")
    );
  }

  // Inverse of b58encode: multiply-and-add across the byte buffer.
  function b58decode(text) {
    let zeros = 0;
    while (zeros < text.length && text[zeros] === "1") zeros++;

    const bytes = [];
    for (let i = zeros; i < text.length; i++) {
      let value = B58_ALPHABET.indexOf(text[i]);
      if (value < 0) throw new Error("invalid base58btc character: " + text[i]);
      for (let j = 0; j < bytes.length; j++) {
        const acc = bytes[j] * 58 + value;
        bytes[j] = acc & 0xff;
        value = acc >> 8;
      }
      while (value > 0) {
        bytes.push(value & 0xff);
        value >>= 8;
      }
    }
    bytes.reverse();
    return Uint8Array.from(new Array(zeros).fill(0).concat(bytes));
  }

  function hexToBytes(hex) {
    if (typeof hex !== "string" || hex.length % 2 !== 0 || /[^0-9a-fA-F]/.test(hex)) {
      throw new Error("not a hex public key: " + hex);
    }
    const out = new Uint8Array(hex.length / 2);
    for (let i = 0; i < out.length; i++) {
      out[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
    }
    return out;
  }

  function bytesToHex(bytes) {
    return Array.from(bytes)
      .map((b) => b.toString(16).padStart(2, "0"))
      .join("");
  }

  // hex Ed25519 public key -> "did:key:z…". Throws on anything that is not a
  // 32-byte key, so a malformed DID is never displayed as if it were real.
  function publicKeyToDidKey(publicKeyHex) {
    const key = hexToBytes(publicKeyHex);
    if (key.length !== PUBLIC_KEY_LEN) {
      throw new Error(
        "expected a " + PUBLIC_KEY_LEN + "-byte Ed25519 public key, got " + key.length
      );
    }
    const payload = Uint8Array.from(ED25519_MULTICODEC_PREFIX.concat(Array.from(key)));
    return DID_KEY_PREFIX + MULTIBASE_BASE58BTC + b58encode(payload);
  }

  // "did:key:z…" -> the raw 32 verification-key bytes (exact inverse).
  function didKeyToPublicKey(did) {
    if (typeof did !== "string" || did.indexOf(DID_KEY_PREFIX) !== 0) {
      throw new Error("not a did:key identifier: " + did);
    }
    const body = did.slice(DID_KEY_PREFIX.length);
    if (body[0] !== MULTIBASE_BASE58BTC) {
      throw new Error("did:key must use the base58btc multibase prefix 'z'");
    }
    const payload = b58decode(body.slice(1));
    if (
      payload[0] !== ED25519_MULTICODEC_PREFIX[0] ||
      payload[1] !== ED25519_MULTICODEC_PREFIX[1]
    ) {
      throw new Error("did:key is not an ed25519-pub key");
    }
    const key = payload.slice(ED25519_MULTICODEC_PREFIX.length);
    if (key.length !== PUBLIC_KEY_LEN) {
      throw new Error("did:key payload is " + key.length + " bytes");
    }
    return key;
  }

  function didKeyToPublicHex(did) {
    return bytesToHex(didKeyToPublicKey(did));
  }

  // Display helper: a DID when the value really is an Ed25519 key, else null.
  // The UI labels placeholders too (the genesis block's absent producer), and
  // those should render without a DID rather than throw mid-render.
  function tryPublicKeyToDidKey(value) {
    try {
      return publicKeyToDidKey(value);
    } catch {
      return null;
    }
  }

  const api = {
    publicKeyToDidKey,
    didKeyToPublicKey,
    didKeyToPublicHex,
    tryPublicKeyToDidKey,
  };

  global.didKey = api;
  // Also reachable from Node (tests/test_did_js.py runs this file directly to
  // check the mirror against the Python implementation).
  if (typeof module !== "undefined" && module.exports) module.exports = api;
})(typeof window !== "undefined" ? window : globalThis);

File: crypto/did.py
"""did:key identities — a *naming* layer over the existing Ed25519 keys.

This module introduces no new cryptography and no new key material. Every
identity in the system is still exactly what it was: a 32-byte Ed25519 public
key, hex-encoded, recorded verbatim in a transaction's ``sender``. What a
``did:key`` adds is a second, standards-shaped *rendering* of that same key —
self-certifying (the key is inside the identifier), resolvable without any
registry, and portable to any wallet or verifier that speaks W3C DIDs.

The encoding is the one the did:key method specifies for Ed25519:

    did:key:z<base58btc( 0xed 0x01 || raw_public_key_bytes )>

- ``0xed 0x01`` is the ``ed25519-pub`` multicodec code (0xed) written as an
  unsigned varint, which is why the low-order byte comes first.
- ``z`` is the multibase prefix for base58btc, so the identifier declares its
  own base rather than leaving it to convention.

Because the key is the whole payload, :func:`did_key_to_public_key` is a pure
inverse: it recovers the same bytes, with no lookup and no trust in anything
but the string. Nothing here touches private keys, hashing, signing, or
consensus — a DID is derived from a public key on demand and never stored.
"""

from __future__ import annotations

# The ed25519-pub multicodec code (0xed) as an unsigned varint: values above
# 0x7f are emitted low-7-bits-first with the continuation bit set on all but
# the last byte, so 0xed becomes 0xed 0x01.
ED25519_MULTICODEC_PREFIX = b"\xed\x01"

# The multibase prefix declaring "the rest of this string is base58btc".
MULTIBASE_BASE58BTC = "z"

DID_KEY_PREFIX = "did:key:"

# Bitcoin's base58 alphabet: the digits and letters minus 0, O, I and l, which
# are the characters people mistake for one another when reading a key aloud.
_B58_ALPHABET = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"
_B58_INDEX = {char: value for value, char in enumerate(_B58_ALPHABET)}

_ED25519_PUBLIC_KEY_LEN = 32


def _b58encode(data: bytes) -> str:
    """base58btc-encode ``data`` (Bitcoin alphabet, leading zeros as ``1``).

    Leading zero bytes carry no weight in the big-integer conversion, so they
    are counted separately and re-emitted as ``1`` characters — that is what
    makes the encoding injective and the decode exactly reversible.
    """
    leading_zeros = len(data) - len(data.lstrip(b"\x00"))
    number = int.from_bytes(data, "big")
    digits = ""
    while number > 0:
        number, remainder = divmod(number, 58)
        digits = _B58_ALPHABET[remainder] + digits
    return "1" * leading_zeros + digits


def _b58decode(text: str) -> bytes:
    """Inverse of :func:`_b58encode`. Raises ``ValueError`` on any stray character."""
    number = 0
    for char in text:
        value = _B58_INDEX.get(char)
        if value is None:
            raise ValueError(f"invalid base58btc character: {char!r}")
        number = number * 58 + value
    leading_zeros = len(text) - len(text.lstrip("1"))
    body = number.to_bytes((number.bit_length() + 7) // 8, "big") if number else b""
    return b"\x00" * leading_zeros + body


def public_key_to_did_key(public_key_hex: str) -> str:
    """Render a hex Ed25519 public key as its ``did:key`` identifier.

    Args:
        public_key_hex: The 64-character hex encoding of a 32-byte Ed25519
            public key — the same string the chain stores in ``sender``.

    Returns:
        ``"did:key:z..."`` — a deterministic function of the key alone, so the
        same key always names the same DID on every node and in the browser.

    Raises:
        ValueError: If the input is not 32 bytes of valid hex. A DID that does
            not decode back to a usable verification key would be worse than no
            DID at all, so this is checked rather than papered over.
    """
    try:
        key_bytes = bytes.fromhex(public_key_hex)
    except ValueError as exc:
        raise ValueError(f"not a hex public key: {public_key_hex!r}") from exc
    if len(key_bytes) != _ED25519_PUBLIC_KEY_LEN:
        raise ValueError(
            f"expected a {_ED25519_PUBLIC_KEY_LEN}-byte Ed25519 public key, "
            f"got {len(key_bytes)} bytes"
        )
    payload = ED25519_MULTICODEC_PREFIX + key_bytes
    return DID_KEY_PREFIX + MULTIBASE_BASE58BTC + _b58encode(payload)


def did_key_to_public_key(did: str) -> bytes:
    """Recover the raw Ed25519 verification key bytes from a ``did:key``.

    The exact inverse of :func:`public_key_to_did_key`: no resolution step and
    no network, because the key *is* the identifier.

    Raises:
        ValueError: If ``did`` is not a base58btc ``did:key`` carrying an
            ``ed25519-pub`` multicodec payload of the right length. Other key
            types are legal did:keys but are not identities this chain can
            verify a signature against, so they are rejected explicitly.
    """
    if not isinstance(did, str) or not did.startswith(DID_KEY_PREFIX):
        raise ValueError(f"not a did:key identifier: {did!r}")
    body = did[len(DID_KEY_PREFIX) :]
    if not body.startswith(MULTIBASE_BASE58BTC):
        raise ValueError("did:key must use the base58btc multibase prefix 'z'")
    payload = _b58decode(body[len(MULTIBASE_BASE58BTC) :])
    if not payload.startswith(ED25519_MULTICODEC_PREFIX):
        raise ValueError("did:key is not an ed25519-pub key")
    key_bytes = payload[len(ED25519_MULTICODEC_PREFIX) :]
    if len(key_bytes) != _ED25519_PUBLIC_KEY_LEN:
        raise ValueError(
            f"did:key payload is {len(key_bytes)} bytes, "
            f"expected {_ED25519_PUBLIC_KEY_LEN}"
        )
    return key_bytes


def did_key_to_public_hex(did: str) -> str:
    """:func:`did_key_to_public_key` in the hex form the rest of the system uses."""
    return did_key_to_public_key(did).hex()


def try_public_key_to_did_key(value: str) -> str | None:
    """Best-effort DID for ``value``, or ``None`` if it is not an Ed25519 key.

    The UI labels things that are not always real keys — the genesis block's
    absent producer, readable placeholder names in tests and fixtures. Those
    should render without a DID rather than fail the whole response, so display
    code calls this instead of the strict function.
    """
    try:
        return public_key_to_did_key(value)
    except (ValueError, TypeError):
        return None
